// @duneanalytics/client-sdk/dist/esm/types/requestArgs.js
import assert from "assert";
import { QueryParameter } from "./queryParameter";
/// Choice of execution engine when executing query via API [default = medium]
export var QueryEngine;
(function (QueryEngine) {
    QueryEngine["Medium"] = "medium";
    QueryEngine["Large"] = "large";
})(QueryEngine || (QueryEngine = {}));
/// Utility method used by router to parse request payloads.
export function payloadJSON(payload) {
    return JSON.stringify(payloadRecords(payload));
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function payloadRecords(payload) {
    if (payload !== undefined) {
        if ("query_parameters" in payload) {
            // Destructure to separate parameters and the rest of the payload
            const { query_parameters, ...rest } = payload;
            return {
                ...rest,
                query_parameters: query_parameters
                    ? QueryParameter.unravel(query_parameters)
                    : [],
            };
        }
        return payload;
    }
    return {};
}
/**
 * Converts all arguments into a format
 * which can be converted into a URL path for GET requests.
 */
export function payloadSearchParams(payload) {
    if (payload !== undefined) {
        const intermPayload = payload;
        if ("query_parameters" in payload) {
            // Destructure to separate parameters and the rest of the payload
            const { query_parameters, ...rest } = intermPayload;
            // Remove all undefined keys from payload.
            const result = Object.keys(rest).reduce((acc, key) => {
                if (rest[key] !== undefined) {
                    acc[key] = rest[key];
                }
                return acc;
            }, {});
            // Modify query parameter to satisfy API formating requirements.
            if (Array.isArray(payload.query_parameters)) {
                for (const qp of payload.query_parameters) {
                    result[`params.${qp.name}`] = qp.value;
                }
            }
            return result;
        }
        return payload;
    }
    return {};
}
export function validateAndBuildGetResultParams({ limit, offset, sample_count, filters, sort_by, columns, query_parameters, }) {
    assert(sample_count === undefined ||
        (limit === undefined && offset === undefined && filters === undefined), "sampling cannot be combined with filters or pagination");
    if (columns !== undefined) {
        if (typeof columns === "string") {
            columns = columns.split(",");
        }
        const output = columns.map((column) => {
            // Check if the column contains quotes
            if (column.includes('"')) {
                // Escape quotes and add quotes around the entire string
                return `"${column.replace(/"/g, '\\"')}"`;
            }
            else {
                // Leave the column unchanged
                return column;
            }
        });
        columns = output.join(",");
    }
    if (sort_by !== undefined && Array.isArray(sort_by)) {
        sort_by = sort_by.join(",");
    }
    query_parameters = query_parameters || [];
    return {
        // It used to be the case that limit was required,
        // but now that they have introduced some other filters that
        // are incompatible with this field, it is no longer required.
        // It is becomes required again later, we will need to use withDefaults here.
        limit,
        offset,
        sample_count,
        filters,
        sort_by,
        columns,
        query_parameters,
    };
}
// https://docs.dune.com/api-reference/tables/endpoint/create#body-schema-type
export var ColumnType;
(function (ColumnType) {
    ColumnType["Varchar"] = "varchar";
    ColumnType["Varbinary"] = "varbinary";
    ColumnType["Uint256"] = "uint256";
    ColumnType["Int256"] = "int256";
    ColumnType["Bigint"] = "bigint";
    ColumnType["Integer"] = "integer";
    ColumnType["Double"] = "double";
    ColumnType["Boolean"] = "boolean";
    ColumnType["Timestamp"] = "timestamp";
    ColumnType["Date"] = "date";
})(ColumnType || (ColumnType = {}));
/**
 * All supported API content types
 */
export var ContentType;
(function (ContentType) {
    ContentType["Json"] = "application/json";
    ContentType["Csv"] = "text/csv";
    ContentType["NDJson"] = "application/x-ndjson";
})(ContentType || (ContentType = {}));
